src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/PostProcessReadsForRSEM.java

```diff
@@ -0,0 +1,257 @@
+)
+public class PostProcessReadsForRSEM extends GATKTool {
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME)
+    public File outSam;
+
+    @Argument(fullName = "keep-MT-reads")
+    public boolean keepMTReads = false;
+
+    PeekableIterator<GATKRead> read1Iterator;
+    SAMFileGATKReadWriter writer;
+
+    ReadPair currentReadPair;
+
+    // Use CountingFilter or another existing framework for counting and printing filtered reads
+    int totalOutputPair = 0;
+    int notBothMapped = 0;
+    int chimera = 0;
+    int unsupportedCigar = 0;
+    int duplicates = 0;
+    int mitochondria = 0;
+
+    @Override
+    public void onTraversalStart(){
+        read1Iterator = new PeekableIterator<>(directlyAccessEngineReadsDataSource().iterator());
+
+        if (directlyAccessEngineReadsDataSource().getHeader().getSortOrder() != SAMFileHeader.SortOrder.queryname){
+            throw new UserException("Input must be query-name sorted.");
+        }
+
+        writer = createSAMWriter(new GATKPath(outSam.getAbsolutePath()), true);
+
+        if (!read1Iterator.hasNext()){
+            throw new UserException("Input has no reads");
+        }
+
+        currentReadPair = new ReadPair(read1Iterator.next());
+        totalOutputPair++;
+    }
+
+    @Override
+    public void traverse() {
+        while (read1Iterator.hasNext()){
+            final GATKRead read = read1Iterator.next();
+            if (!currentReadPair.getQueryName().equals(read.getName())){
+                // We have gathered all the reads with the same query name (primary, secondary, and supplementary alignments)
+                // Write the reads to output file, reordering the reads as needed.
+                writeReads(currentReadPair);
+                currentReadPair = new ReadPair(read);
+            } else {
+                currentReadPair.add(read);
+            }
+            progressMeter.update(read);
+        }
+    }
+
+    public boolean passesRSEMFilter(final GATKRead read1, final GATKRead read2) {
+        // If either of the pair is unmapped, throw out.
+        // With the STAR argument --quantTranscriptomeBan IndelSoftclipSingleend this should not occur,
+        // but we check just to be thorough.
+        if (read1.getContig() == null || read2.getContig() == null){
+            notBothMapped++;
+            return false;
+        }
+
+        // Chimeric reads are not allowed
+        if (!read1.getContig().equals(read2.getContig())){
+            chimera++;
+            return false;
+        }
+
+        // Cigar must have a single operator M.
+        final List<CigarElement> cigarElements1 = read1.getCigar().getCigarElements();
+        final List<CigarElement> cigarElements2 = read2.getCigar().getCigarElements();
+
+        if (cigarElements1.size() != 1 || cigarElements2.size() != 1){
```

Comment:
I might pull out a little function and call it twice instead of duplicating the checks.